{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3cc45e64",
   "metadata": {},
   "outputs": [],
   "source": [
    "%load_ext autoreload\n",
    "%autoreload 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "2d35f5c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import sys\n",
    "\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "\n",
    "sys.path.append(\"..\")\n",
    "from utils import read_aws_csv, read_aws_h5ad"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f0a90a7f-b171-48e1-83c9-ab73ec9af158",
   "metadata": {},
   "source": [
    "## Data Preprocessing\n",
    "- Ensure that the data is downloaded (see `../data` for instructions)\n",
    "- Ensure that the data was filtered using `norman_filtering.ipynb`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "b5f89eea",
   "metadata": {},
   "outputs": [],
   "source": [
    "data_path = \"path to filtered norman.h5ad\"\n",
    "adata = read_aws_h5ad(data_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "cb93a4f5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "105"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# get perturbation genes\n",
    "pert_list = set()\n",
    "for t in adata.obs[\"perturbation_name\"]:\n",
    "    if \"+\" not in t and t != \"control\":\n",
    "        pert_list.add(t)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "c0cc203f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# subset to hvg\n",
    "sc.pp.highly_variable_genes(adata, n_top_genes=5000, subset=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "37d7ff87",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 101578 × 5035\n",
       "    obs: 'guide_identity', 'read_count', 'UMI_count', 'coverage', 'gemgroup', 'good_coverage', 'number_of_cells', 'guide_AHR', 'guide_ARID1A', 'guide_ARRDC3', 'guide_ATL1', 'guide_BAK1', 'guide_BCL2L11', 'guide_BCORL1', 'guide_BPGM', 'guide_C19orf26', 'guide_C3orf72', 'guide_CBFA2T3', 'guide_CBL', 'guide_CDKN1A', 'guide_CDKN1B', 'guide_CDKN1C', 'guide_CEBPA', 'guide_CEBPB', 'guide_CEBPE', 'guide_CELF2', 'guide_CITED1', 'guide_CKS1B', 'guide_CLDN6', 'guide_CNN1', 'guide_CNNM4', 'guide_COL1A1', 'guide_COL2A1', 'guide_CSRNP1', 'guide_DLX2', 'guide_DUSP9', 'guide_EGR1', 'guide_ELMSAN1', 'guide_ETS2', 'guide_FEV', 'guide_FOSB', 'guide_FOXA1', 'guide_FOXA3', 'guide_FOXF1', 'guide_FOXL2', 'guide_FOXO4', 'guide_GLB1L2', 'guide_HES7', 'guide_HK2', 'guide_HNF4A', 'guide_HOXA13', 'guide_HOXB9', 'guide_HOXC13', 'guide_IER5L', 'guide_IGDCC3', 'guide_IKZF3', 'guide_IRF1', 'guide_ISL2', 'guide_JUN', 'guide_KIAA1804', 'guide_KIF18B', 'guide_KIF2C', 'guide_KLF1', 'guide_KMT2A', 'guide_LHX1', 'guide_LYL1', 'guide_MAML2', 'guide_MAP2K3', 'guide_MAP2K6', 'guide_MAP4K3', 'guide_MAP4K5', 'guide_MAP7D1', 'guide_MAPK1', 'guide_MEIS1', 'guide_MIDN', 'guide_NCL', 'guide_NIT1', 'guide_OSR2', 'guide_PLK4', 'guide_POU3F2', 'guide_PRDM1', 'guide_PRTG', 'guide_PTPN1', 'guide_PTPN12', 'guide_PTPN13', 'guide_PTPN9', 'guide_RHOXF2', 'guide_RREB1', 'guide_RUNX1T1', 'guide_S1PR2', 'guide_SAMD1', 'guide_SET', 'guide_SGK1', 'guide_SLC38A2', 'guide_SLC4A1', 'guide_SLC6A9', 'guide_SNAI1', 'guide_SPI1', 'guide_STIL', 'guide_TBX2', 'guide_TBX3', 'guide_TGFBR2', 'guide_TMSB4X', 'guide_TP73', 'guide_TSC22D1', 'guide_UBASH3A', 'guide_UBASH3B', 'guide_ZBTB1', 'guide_ZBTB10', 'guide_ZBTB25', 'guide_ZC3HAV1', 'guide_ZNF318', 'guide_ids', 'n_genes', 'n_genes_by_counts', 'total_counts', 'total_counts_mt', 'pct_counts_mt', 'leiden', 'perturbation_name', 'perturbation_type', 'perturbation_value', 'perturbation_unit', 'log1p_n_genes_by_counts', 'log1p_total_counts'\n",
       "    var: 'index', 'n_cells', 'mt', 'n_cells_by_counts', 'mean_counts', 'pct_dropout_by_counts', 'total_counts', 'highly_variable', 'means', 'dispersions', 'dispersions_norm', 'log1p_mean_counts', 'log1p_total_counts'\n",
       "    uns: 'doi', 'guide_ids_colors', 'hvg', 'leiden', 'leiden_colors', 'neighbors', 'pca', 'perturbation_type_colors', 'preprocessing_nb_link', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'\n",
       "    layers: 'counts'\n",
       "    obsp: 'connectivities', 'distances'"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# ensure all perts are in adata\n",
    "idx = []\n",
    "for i in range(len(adata.var_names)):\n",
    "    if (adata.var[\"highly_variable\"][i]) or (adata.var_names[i] in pert_list):\n",
    "        idx.append(i)\n",
    "adata_subset = adata[:, idx]\n",
    "adata_subset"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "99cd823e-a774-4e07-a6b1-d1586e5411e3",
   "metadata": {},
   "source": [
    "## StringDB prior graph"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "cade1ddc",
   "metadata": {},
   "outputs": [],
   "source": [
    "# stringdb prior\n",
    "stringdb_hq = read_aws_csv(\"s3://pert-spectra/references/StringDB.HQ.txt\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "9410883b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# ensg mapping\n",
    "gene_name_df = read_aws_csv(\n",
    "    \"s3://pert-spectra/references/Homo_sapiens.gene_info\", sep=\"\\t\"\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "977803fa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#tax_id</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>Symbol</th>\n",
       "      <th>LocusTag</th>\n",
       "      <th>Synonyms</th>\n",
       "      <th>dbXrefs</th>\n",
       "      <th>chromosome</th>\n",
       "      <th>map_location</th>\n",
       "      <th>description</th>\n",
       "      <th>type_of_gene</th>\n",
       "      <th>Symbol_from_nomenclature_authority</th>\n",
       "      <th>Full_name_from_nomenclature_authority</th>\n",
       "      <th>Nomenclature_status</th>\n",
       "      <th>Other_designations</th>\n",
       "      <th>Modification_date</th>\n",
       "      <th>Feature_type</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>9606</td>\n",
       "      <td>1</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>-</td>\n",
       "      <td>A1B|ABG|GAB|HYST2477</td>\n",
       "      <td>MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410...</td>\n",
       "      <td>19</td>\n",
       "      <td>19q13.43</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A1BG</td>\n",
       "      <td>alpha-1-B glycoprotein</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...</td>\n",
       "      <td>20240617</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>9606</td>\n",
       "      <td>2</td>\n",
       "      <td>A2M</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MD|CPAMD5|FWP007|S863-7</td>\n",
       "      <td>MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899...</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>A2M</td>\n",
       "      <td>alpha-2-macroglobulin</td>\n",
       "      <td>O</td>\n",
       "      <td>alpha-2-macroglobulin|C3 and PZP-like alpha-2-...</td>\n",
       "      <td>20240617</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>9606</td>\n",
       "      <td>3</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>-</td>\n",
       "      <td>A2MP</td>\n",
       "      <td>HGNC:HGNC:8|Ensembl:ENSG00000291190|AllianceGe...</td>\n",
       "      <td>12</td>\n",
       "      <td>12p13.31</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>A2MP1</td>\n",
       "      <td>alpha-2-macroglobulin pseudogene 1</td>\n",
       "      <td>O</td>\n",
       "      <td>pregnancy-zone protein pseudogene</td>\n",
       "      <td>20240617</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>9606</td>\n",
       "      <td>9</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC1|MNAT|NAT-1|NATI</td>\n",
       "      <td>MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171...</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT1</td>\n",
       "      <td>N-acetyltransferase 1</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 1|N-acetyltransf...</td>\n",
       "      <td>20240617</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>9606</td>\n",
       "      <td>10</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>-</td>\n",
       "      <td>AAC2|NAT-2|PNAT</td>\n",
       "      <td>MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156...</td>\n",
       "      <td>8</td>\n",
       "      <td>8p22</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>NAT2</td>\n",
       "      <td>N-acetyltransferase 2</td>\n",
       "      <td>O</td>\n",
       "      <td>arylamine N-acetyltransferase 2|N-acetyltransf...</td>\n",
       "      <td>20240617</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193451</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923215</td>\n",
       "      <td>trnD</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>tRNA-Asp</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193452</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923216</td>\n",
       "      <td>trnP</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>tRNA-Pro</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193453</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923217</td>\n",
       "      <td>trnA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>tRNA-Ala</td>\n",
       "      <td>tRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193454</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923218</td>\n",
       "      <td>COX1</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>cytochrome c oxidase subunit I</td>\n",
       "      <td>protein-coding</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>cytochrome c oxidase subunit I</td>\n",
       "      <td>20230818</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>193455</th>\n",
       "      <td>741158</td>\n",
       "      <td>8923219</td>\n",
       "      <td>16S rRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>MT</td>\n",
       "      <td>-</td>\n",
       "      <td>l-rRNA</td>\n",
       "      <td>rRNA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>20200909</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>193456 rows × 16 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        #tax_id   GeneID    Symbol LocusTag                   Synonyms  \\\n",
       "0          9606        1      A1BG        -       A1B|ABG|GAB|HYST2477   \n",
       "1          9606        2       A2M        -  A2MD|CPAMD5|FWP007|S863-7   \n",
       "2          9606        3     A2MP1        -                       A2MP   \n",
       "3          9606        9      NAT1        -       AAC1|MNAT|NAT-1|NATI   \n",
       "4          9606       10      NAT2        -            AAC2|NAT-2|PNAT   \n",
       "...         ...      ...       ...      ...                        ...   \n",
       "193451   741158  8923215      trnD        -                          -   \n",
       "193452   741158  8923216      trnP        -                          -   \n",
       "193453   741158  8923217      trnA        -                          -   \n",
       "193454   741158  8923218      COX1        -                          -   \n",
       "193455   741158  8923219  16S rRNA        -                          -   \n",
       "\n",
       "                                                  dbXrefs chromosome  \\\n",
       "0       MIM:138670|HGNC:HGNC:5|Ensembl:ENSG00000121410...         19   \n",
       "1       MIM:103950|HGNC:HGNC:7|Ensembl:ENSG00000175899...         12   \n",
       "2       HGNC:HGNC:8|Ensembl:ENSG00000291190|AllianceGe...         12   \n",
       "3       MIM:108345|HGNC:HGNC:7645|Ensembl:ENSG00000171...          8   \n",
       "4       MIM:612182|HGNC:HGNC:7646|Ensembl:ENSG00000156...          8   \n",
       "...                                                   ...        ...   \n",
       "193451                                                  -         MT   \n",
       "193452                                                  -         MT   \n",
       "193453                                                  -         MT   \n",
       "193454                                                  -         MT   \n",
       "193455                                                  -         MT   \n",
       "\n",
       "       map_location                         description    type_of_gene  \\\n",
       "0          19q13.43              alpha-1-B glycoprotein  protein-coding   \n",
       "1          12p13.31               alpha-2-macroglobulin  protein-coding   \n",
       "2          12p13.31  alpha-2-macroglobulin pseudogene 1          pseudo   \n",
       "3              8p22               N-acetyltransferase 1  protein-coding   \n",
       "4              8p22               N-acetyltransferase 2  protein-coding   \n",
       "...             ...                                 ...             ...   \n",
       "193451            -                            tRNA-Asp            tRNA   \n",
       "193452            -                            tRNA-Pro            tRNA   \n",
       "193453            -                            tRNA-Ala            tRNA   \n",
       "193454            -      cytochrome c oxidase subunit I  protein-coding   \n",
       "193455            -                              l-rRNA            rRNA   \n",
       "\n",
       "       Symbol_from_nomenclature_authority  \\\n",
       "0                                    A1BG   \n",
       "1                                     A2M   \n",
       "2                                   A2MP1   \n",
       "3                                    NAT1   \n",
       "4                                    NAT2   \n",
       "...                                   ...   \n",
       "193451                                  -   \n",
       "193452                                  -   \n",
       "193453                                  -   \n",
       "193454                                  -   \n",
       "193455                                  -   \n",
       "\n",
       "       Full_name_from_nomenclature_authority Nomenclature_status  \\\n",
       "0                     alpha-1-B glycoprotein                   O   \n",
       "1                      alpha-2-macroglobulin                   O   \n",
       "2         alpha-2-macroglobulin pseudogene 1                   O   \n",
       "3                      N-acetyltransferase 1                   O   \n",
       "4                      N-acetyltransferase 2                   O   \n",
       "...                                      ...                 ...   \n",
       "193451                                     -                   -   \n",
       "193452                                     -                   -   \n",
       "193453                                     -                   -   \n",
       "193454                                     -                   -   \n",
       "193455                                     -                   -   \n",
       "\n",
       "                                       Other_designations  Modification_date  \\\n",
       "0       alpha-1B-glycoprotein|HEL-S-163pA|epididymis s...           20240617   \n",
       "1       alpha-2-macroglobulin|C3 and PZP-like alpha-2-...           20240617   \n",
       "2                       pregnancy-zone protein pseudogene           20240617   \n",
       "3       arylamine N-acetyltransferase 1|N-acetyltransf...           20240617   \n",
       "4       arylamine N-acetyltransferase 2|N-acetyltransf...           20240617   \n",
       "...                                                   ...                ...   \n",
       "193451                                                  -           20200909   \n",
       "193452                                                  -           20200909   \n",
       "193453                                                  -           20200909   \n",
       "193454                     cytochrome c oxidase subunit I           20230818   \n",
       "193455                                                  -           20200909   \n",
       "\n",
       "       Feature_type  \n",
       "0                 -  \n",
       "1                 -  \n",
       "2                 -  \n",
       "3                 -  \n",
       "4                 -  \n",
       "...             ...  \n",
       "193451            -  \n",
       "193452            -  \n",
       "193453            -  \n",
       "193454            -  \n",
       "193455            -  \n",
       "\n",
       "[193456 rows x 16 columns]"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_name_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "837bd514",
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_to_ensg = {}\n",
    "gene_name_mapping = {}\n",
    "for symbol, synonyms, ref in zip(\n",
    "    gene_name_df[\"Symbol\"], gene_name_df[\"Synonyms\"], gene_name_df[\"dbXrefs\"]\n",
    "):\n",
    "    syn_set = set(synonyms.split(\"|\")).union(set([symbol]))\n",
    "    refs = ref.split(\"|\")\n",
    "    ensg = None\n",
    "    for r in refs:\n",
    "        label = r.split(\":\")\n",
    "        if label[0] == \"Ensembl\":\n",
    "            ensg = label[1]\n",
    "            break\n",
    "    if ensg is None:\n",
    "        continue\n",
    "    assert ensg[:4] == \"ENSG\"\n",
    "    gene_to_ensg[symbol] = ensg\n",
    "    gene_name_mapping[symbol] = syn_set\n",
    "    # make sure name mapping goes both ways\n",
    "    for syn in syn_set:\n",
    "        if syn in gene_name_mapping:\n",
    "            gene_name_mapping[syn].add(symbol)\n",
    "        else:\n",
    "            gene_name_mapping[syn] = set([symbol])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "4c10530a",
   "metadata": {},
   "outputs": [],
   "source": [
    "adata_subset.var[\"gene_symbols\"] = adata_subset.var_names.copy()\n",
    "adata_subset.var_names = adata_subset.var[\"index\"].copy()\n",
    "norman_ensg_to_gene = {}\n",
    "for symbol, ensgid in zip(adata_subset.var[\"gene_symbols\"], adata_subset.var_names):\n",
    "    norman_ensg_to_gene[ensgid] = symbol"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "id": "5a884a01",
   "metadata": {},
   "outputs": [],
   "source": [
    "dataset_measured_genes = set(adata_subset.var[\"gene_symbols\"])\n",
    "dataset_measured_ensg = set(adata_subset.var_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "id": "73bd2fda",
   "metadata": {},
   "outputs": [],
   "source": [
    "# first, check if gene name has ensembl id\n",
    "# if it does, use that to match\n",
    "# else, use gene name mapping\n",
    "edges = []\n",
    "for index, row in stringdb_hq.iterrows():\n",
    "    gene1 = row[\"i_genes\"]\n",
    "    gene2 = row[\"j_genes\"]\n",
    "    if gene1 not in gene_to_ensg:\n",
    "        # neither gene has ensg\n",
    "        if gene2 not in gene_to_ensg:\n",
    "            gene1_syn = set([gene1])\n",
    "            gene2_syn = set([gene2])\n",
    "            if gene1 in gene_name_mapping:\n",
    "                gene1_syn = gene_name_mapping[gene1].union(set([gene1]))\n",
    "            if gene2 in gene_name_mapping:\n",
    "                gene2_syn = gene_name_mapping[gene2].union(set([gene2]))\n",
    "\n",
    "            alias_intersect_1 = gene1_syn.intersection(dataset_measured_genes)\n",
    "            alias_intersect_2 = gene2_syn.intersection(dataset_measured_genes)\n",
    "\n",
    "            if (len(alias_intersect_1) > 0) and (len(alias_intersect_2) > 0):\n",
    "                gene1 = (\n",
    "                    gene1\n",
    "                    if gene1 in dataset_measured_genes\n",
    "                    else list(alias_intersect_1)[0]\n",
    "                )\n",
    "                gene2 = (\n",
    "                    gene2\n",
    "                    if gene2 in dataset_measured_genes\n",
    "                    else list(alias_intersect_2)[0]\n",
    "                )\n",
    "                edges.append([gene1, gene2, row[\"x\"]])\n",
    "\n",
    "        # gene1 does not have ensemblid, gene2 does have ensemblid\n",
    "        else:\n",
    "            ensg2 = gene_to_ensg[gene2]\n",
    "            gene1_syn = set([gene1])\n",
    "            gene2_syn = set([ensg2])\n",
    "            if gene1 in gene_name_mapping:\n",
    "                gene1_syn = gene_name_mapping[gene1].union(set([gene1]))\n",
    "\n",
    "            alias_intersect_1 = gene1_syn.intersection(dataset_measured_genes)\n",
    "            alias_intersect_2 = gene2_syn.intersection(dataset_measured_ensg)\n",
    "\n",
    "            if (len(alias_intersect_1) > 0) and (len(alias_intersect_2) > 0):\n",
    "                gene1 = (\n",
    "                    gene1\n",
    "                    if gene1 in dataset_measured_genes\n",
    "                    else list(alias_intersect_1)[0]\n",
    "                )\n",
    "                gene2 = norman_ensg_to_gene[list(alias_intersect_2)[0]]\n",
    "                edges.append([gene1, gene2, row[\"x\"]])\n",
    "\n",
    "    else:\n",
    "        # gene1 has ensemblid, gene2 does not have ensemblid\n",
    "        if gene2 not in gene_to_ensg:\n",
    "            ensg1 = gene_to_ensg[gene1]\n",
    "            gene1_syn = set([ensg1])\n",
    "            gene2_syn = set([gene2])\n",
    "            if gene2 in gene_name_mapping:\n",
    "                gene2_syn = gene_name_mapping[gene2].union(set([gene2]))\n",
    "\n",
    "            alias_intersect_1 = gene1_syn.intersection(dataset_measured_ensg)\n",
    "            alias_intersect_2 = gene2_syn.intersection(dataset_measured_genes)\n",
    "\n",
    "            if (len(alias_intersect_1) > 0) and (len(alias_intersect_2) > 0):\n",
    "                gene1 = norman_ensg_to_gene[list(alias_intersect_1)[0]]\n",
    "                gene2 = (\n",
    "                    gene2\n",
    "                    if gene2 in dataset_measured_genes\n",
    "                    else list(alias_intersect_2)[0]\n",
    "                )\n",
    "                edges.append([gene1, gene2, row[\"x\"]])\n",
    "\n",
    "        # both genes have ensmblid\n",
    "        else:\n",
    "            ensg1 = gene_to_ensg[gene1]\n",
    "            ensg2 = gene_to_ensg[gene2]\n",
    "            gene1_syn = set([ensg1])\n",
    "            gene2_syn = set([ensg2])\n",
    "\n",
    "            alias_intersect_1 = gene1_syn.intersection(dataset_measured_ensg)\n",
    "            alias_intersect_2 = gene2_syn.intersection(dataset_measured_ensg)\n",
    "\n",
    "            if (len(alias_intersect_1) > 0) and (len(alias_intersect_2) > 0):\n",
    "                gene1 = norman_ensg_to_gene[list(alias_intersect_1)[0]]\n",
    "                gene2 = norman_ensg_to_gene[list(alias_intersect_2)[0]]\n",
    "                edges.append([gene1, gene2, row[\"x\"]])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "id": "9e2e5c9c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MISP</td>\n",
       "      <td>USH1C</td>\n",
       "      <td>0.212</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>TUBB1</td>\n",
       "      <td>USH1C</td>\n",
       "      <td>0.184</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CRYM</td>\n",
       "      <td>USH1C</td>\n",
       "      <td>0.289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>LIN7B</td>\n",
       "      <td>USH1C</td>\n",
       "      <td>0.152</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>TAX1BP3</td>\n",
       "      <td>USH1C</td>\n",
       "      <td>0.919</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223491</th>\n",
       "      <td>N4BP2</td>\n",
       "      <td>ZNF518A</td>\n",
       "      <td>0.179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223492</th>\n",
       "      <td>PHIP</td>\n",
       "      <td>ZNF518A</td>\n",
       "      <td>0.160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223493</th>\n",
       "      <td>BRD2</td>\n",
       "      <td>ZNF518A</td>\n",
       "      <td>0.228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223494</th>\n",
       "      <td>GSTP1</td>\n",
       "      <td>ZNF518A</td>\n",
       "      <td>0.165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223495</th>\n",
       "      <td>MYL12B</td>\n",
       "      <td>ZNF518A</td>\n",
       "      <td>0.189</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>223496 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene1    gene2  score\n",
       "0          MISP    USH1C  0.212\n",
       "1         TUBB1    USH1C  0.184\n",
       "2          CRYM    USH1C  0.289\n",
       "3         LIN7B    USH1C  0.152\n",
       "4       TAX1BP3    USH1C  0.919\n",
       "...         ...      ...    ...\n",
       "223491    N4BP2  ZNF518A  0.179\n",
       "223492     PHIP  ZNF518A  0.160\n",
       "223493     BRD2  ZNF518A  0.228\n",
       "223494    GSTP1  ZNF518A  0.165\n",
       "223495   MYL12B  ZNF518A  0.189\n",
       "\n",
       "[223496 rows x 3 columns]"
      ]
     },
     "execution_count": 74,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "norman_network = pd.DataFrame(edges, columns=[\"gene1\", \"gene2\", \"score\"])\n",
    "norman_network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "id": "b6bd6d19",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene1</th>\n",
       "      <th>gene2</th>\n",
       "      <th>score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>ZNF720</td>\n",
       "      <td>ZNF720</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>LINC00467</td>\n",
       "      <td>LINC00467</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>RP3-473L9.4</td>\n",
       "      <td>RP3-473L9.4</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>RP11-380O24.1</td>\n",
       "      <td>RP11-380O24.1</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>LINC01597</td>\n",
       "      <td>LINC01597</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225103</th>\n",
       "      <td>N4BP2</td>\n",
       "      <td>ZNF518A</td>\n",
       "      <td>0.179</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225104</th>\n",
       "      <td>PHIP</td>\n",
       "      <td>ZNF518A</td>\n",
       "      <td>0.160</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225105</th>\n",
       "      <td>BRD2</td>\n",
       "      <td>ZNF518A</td>\n",
       "      <td>0.228</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225106</th>\n",
       "      <td>GSTP1</td>\n",
       "      <td>ZNF518A</td>\n",
       "      <td>0.165</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>225107</th>\n",
       "      <td>MYL12B</td>\n",
       "      <td>ZNF518A</td>\n",
       "      <td>0.189</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>225108 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                gene1          gene2  score\n",
       "0              ZNF720         ZNF720  0.000\n",
       "1           LINC00467      LINC00467  0.000\n",
       "2         RP3-473L9.4    RP3-473L9.4  0.000\n",
       "3       RP11-380O24.1  RP11-380O24.1  0.000\n",
       "4           LINC01597      LINC01597  0.000\n",
       "...               ...            ...    ...\n",
       "225103          N4BP2        ZNF518A  0.179\n",
       "225104           PHIP        ZNF518A  0.160\n",
       "225105           BRD2        ZNF518A  0.228\n",
       "225106          GSTP1        ZNF518A  0.165\n",
       "225107         MYL12B        ZNF518A  0.189\n",
       "\n",
       "[225108 rows x 3 columns]"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# add genes measured in Norman with no connections found in StringDB\n",
    "missing_genes = dataset_measured_genes.difference(set(norman_network[\"gene1\"].unique()))\n",
    "for g in missing_genes:\n",
    "    norman_network = pd.concat(\n",
    "        [pd.DataFrame([[g, g, 0]], columns=norman_network.columns), norman_network],\n",
    "        ignore_index=True,\n",
    "    )\n",
    "norman_network"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "id": "05005170",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create one-hot-encoding mapping based on adata var structure\n",
    "adata_subset.var_names = adata_subset.var[\"gene_symbols\"]\n",
    "\n",
    "\n",
    "def map_gene_to_onehot(name: str):\n",
    "    return adata_subset.var_names.get_loc(name)\n",
    "\n",
    "\n",
    "norman_network[\"gene1\"] = norman_network[\"gene1\"].apply(map_gene_to_onehot)\n",
    "norman_network[\"gene2\"] = norman_network[\"gene2\"].apply(map_gene_to_onehot)\n",
    "norman_network = norman_network.sort_values(by=[\"gene1\"])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "id": "b884a780",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create weighted adj matrix\n",
    "import networkx\n",
    "\n",
    "edgeList = norman_network.values.tolist()\n",
    "G = networkx.DiGraph()\n",
    "for i in range(len(edgeList)):\n",
    "    G.add_edge(edgeList[i][0], edgeList[i][1], weight=edgeList[i][2])\n",
    "A = networkx.adjacency_matrix(\n",
    "    G, nodelist=[i for i in range(len(adata_subset.var_names))]\n",
    ").toarray()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "id": "cbd774ea",
   "metadata": {},
   "outputs": [],
   "source": [
    "# turn to sparse\n",
    "from scipy import sparse\n",
    "\n",
    "sA = sparse.csr_matrix(A)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "3f262408",
   "metadata": {},
   "outputs": [],
   "source": [
    "# write to adata\n",
    "adata_subset.uns[\"sparse_gene_network\"] = sA\n",
    "adata_subset.write_h5ad(\"../norman_adata_spectra.h5ad\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "93494551",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "pertspectra",
   "language": "python",
   "name": "pertspectra"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
